File: lib/capsulecd/base/source/github.rb
require 'octokit'
require 'uri'
require 'git'
require 'capsulecd'
require 'pp'

module CapsuleCD
  module Source
    module Github
      # all of these instance variables are available for use within hooks
      attr_accessor :source_client
      attr_accessor :source_git_base_info
      attr_accessor :source_git_head_info
      attr_accessor :source_git_parent_path
      attr_accessor :source_git_local_path
      attr_accessor :source_git_local_branch
      attr_accessor :source_git_remote
      attr_accessor :source_release_commit
      attr_accessor :source_release_artifacts

      # define the Source API methods

      # configure method will generate an authenticated client that can be used to comunicate with Github
      # MUST set @source_git_parent_path
      # MUST set @source_client
      def source_configure
        puts 'github source_configure'
        fail CapsuleCD::Error::SourceAuthenticationFailed, 'Missing "CAPSULE_SOURCE_GITHUB_ACCESS_TOKEN" env variable' unless ENV['CAPSULE_SOURCE_GITHUB_ACCESS_TOKEN']

        @source_release_commit = nil
        @source_release_artifacts = []

        @source_git_parent_path = Dir.mktmpdir
        Octokit.auto_paginate = true
        @source_client = Octokit::Client.new(access_token: ENV['CAPSULE_SOURCE_GITHUB_ACCESS_TOKEN'])
      end

      # all capsule CD processing will be kicked off via a payload. In Github's case, the payload is the webhook data.
      # should check if the pull request opener even has permissions to create a release.
      # all sources should process the payload by downloading a git repository that contains the master branch merged with the test branch
      # MUST set source_git_local_path
      # MUST set source_git_local_branch
      # MUST set source_git_head_info
      # REQUIRES source_git_parent_path
      def source_process_push_payload(payload)
        # set the processed head info
        @source_git_head_info = payload['head']
        CapsuleCD::ValidationUtils.validate_repo_payload(@source_git_head_info)

        # set the remote url, with embedded token
        uri = URI.parse(@source_git_head_info['repo']['clone_url'])
        uri.user = ENV['CAPSULE_SOURCE_GITHUB_ACCESS_TOKEN']
        @source_git_remote = uri.to_s
        @source_git_local_branch = @source_git_head_info['repo']['ref']
        # clone the merged branch
        # https://sethvargo.com/checkout-a-github-pull-request/
        # https://coderwall.com/p/z5rkga/github-checkout-a-pull-request-as-a-branch
        @source_git_local_path = CapsuleCD::GitUtils.clone(@source_git_parent_path, @source_git_head_info['repo']['name'], @source_git_remote)
        CapsuleCD::GitUtils.checkout(@source_git_local_path, @source_git_head_info['repo']['sha1'])
      end

      # all capsule CD processing will be kicked off via a payload. In Github's case, the payload is the pull request data.
      # should check if the pull request opener even has permissions to create a release.
      # all sources should process the payload by downloading a git repository that contains the master branch merged with the test branch
      # MUST set source_git_local_path
      # MUST set source_git_local_branch
      # MUST set source_git_base_info
      # MUST set source_git_head_info
      # REQUIRES source_client
      # REQUIRES source_git_parent_path
      def source_process_pull_request_payload(payload)
        puts 'github source_process_payload'

        # validate the github specific payload options
        unless (payload['state'] == 'open')
          fail CapsuleCD::Error::SourcePayloadUnsupported, 'Pull request has an invalid action'
        end
        unless (payload['base']['repo']['default_branch'] == payload['base']['ref'])
          fail CapsuleCD::Error::SourcePayloadUnsupported, 'Pull request is not being created against the default branch of this repository (usually master)'
        end
        # check the payload push user.

        unless @source_client.collaborator?(payload['base']['repo']['full_name'], payload['user']['login'])

          @source_client.add_comment(payload['base']['repo']['full_name'], payload['number'], CapsuleCD::BotUtils.pull_request_comment)
          fail CapsuleCD::Error::SourceUnauthorizedUser, 'Pull request was opened by an unauthorized user'
        end

        # set the processed base/head info,
        @source_git_base_info = payload['base']
        @source_git_head_info = payload['head']
        CapsuleCD::ValidationUtils.validate_repo_payload(@source_git_base_info)
        CapsuleCD::ValidationUtils.validate_repo_payload(@source_git_head_info)

        # set the remote url, with embedded token
        uri = URI.parse(payload['base']['repo']['clone_url'])
        uri.user = ENV['CAPSULE_SOURCE_GITHUB_ACCESS_TOKEN']
        @source_git_remote = uri.to_s

        # clone the merged branch
        # https://sethvargo.com/checkout-a-github-pull-request/
        # https://coderwall.com/p/z5rkga/github-checkout-a-pull-request-as-a-branch
        @source_git_local_path = CapsuleCD::GitUtils.clone(@source_git_parent_path, @source_git_head_info['repo']['name'], @source_git_remote)
        @source_git_local_branch = "pr_#{payload['number']}"
        CapsuleCD::GitUtils.fetch(@source_git_local_path, "refs/pull/#{payload['number']}/head", @source_git_local_branch)
        CapsuleCD::GitUtils.checkout(@source_git_local_path, @source_git_local_branch)

        # show a processing message on the github PR.
        @source_client.create_status(payload['base']['repo']['full_name'], @source_git_head_info['sha'], 'pending',
                                     target_url: 'http://www.github.com/AnalogJ/capsulecd',
                                     description: 'CapsuleCD has started processing cookbook. Pull request will be merged automatically when complete.')
      end

      # REQUIRES source_client
      # REQUIRES source_release_commit
      # REQUIRES source_git_local_path
      # REQUIRES source_git_local_branch
      # REQUIRES source_git_base_info
      # REQUIRES source_git_head_info
      # REQUIRES source_release_artifacts
      # REQUIRES source_git_parent_path
      def source_release
        puts 'github source_release'

        # push the version bumped metadata file + newly created files to
        CapsuleCD::GitUtils.push(@source_git_local_path, @source_git_local_branch, @source_git_base_info['ref'])
        # sleept because github needs time to process the new tag.
        sleep 5

        # calculate the release sha
        release_sha = ('0' * (40 - @source_release_commit.sha.strip.length)) + @source_release_commit.sha.strip

        # get the release changelog
        release_body = CapsuleCD::GitUtils.generate_changelog(@source_git_local_path, @source_git_base_info['sha'], @source_git_head_info['sha'], @source_git_base_info['repo']['full_name'])

        release = @source_client.create_release(@source_git_base_info['repo']['full_name'], @source_release_commit.name,       target_commitish: release_sha,
                                                                                                                               name: @source_release_commit.name,
                                                                                                                               body: release_body)

        @source_release_artifacts.each do |release_artifact|
          @source_client.upload_asset(release[:url], release_artifact[:path], name: release_artifact[:name])
        end

        FileUtils.remove_entry_secure @source_git_parent_path
        # set the pull request status
        @source_client.create_status(@source_git_base_info['repo']['full_name'], @source_git_head_info['sha'], 'success',      target_url: 'http://www.github.com/AnalogJ/capsulecd',
                                                                                                                               description: 'pull-request was successfully merged, new release created.')
      end

      def source_process_failure(ex)
        puts 'github source_process_failure'
        FileUtils.remove_entry_secure @source_git_parent_path
        @source_client.create_status(@source_git_base_info['repo']['full_name'], @source_git_head_info['sha'], 'failure',       target_url: 'http://www.github.com/AnalogJ/capsulecd',
                                                                                                                                description: ex.message.slice!(0..135))
      end
    end
  end
end

File: lib/scm/github.go
package scm

import (
	"github.com/google/go-github/github"
	"context"
	"golang.org/x/oauth2"
	"capsulecd/lib/config"
	"log"
	"os"
	"io/ioutil"
	"strings"
	"net/http"
	"capsulecd/lib/errors"
	"fmt"
	"net/url"
	"capsulecd/lib/utils"
)

type scmGithub struct {
	options *ScmOptions
	client  *github.Client
}

// configure method will generate an authenticated client that can be used to comunicate with Github
// MUST set options.GitParentPath
// MUST set client
func (g *scmGithub) Configure(client *http.Client) (error) {

	g.options = new(ScmOptions)

	if !config.IsSet("scm_github_access_token") {
		return errors.ScmAuthenticationFailed("Missing github access token")
	}
	if config.IsSet("scm_git_parent_path") {
		g.options.GitParentPath = config.GetString("scm_git_parent_path")
		os.MkdirAll(g.options.GitParentPath, os.ModePerm)
	} else {
		dirPath, err := ioutil.TempDir("","")
		if err != nil {
			return err
		}
		g.options.GitParentPath = dirPath
	}

	if(client != nil){
		//primarily used for testing.
		g.client = github.NewClient(client)
	} else {
		ctx := context.Background()
		ts := oauth2.StaticTokenSource(
			&oauth2.Token{AccessToken: config.GetString("scm_github_access_token")},
		)
		tc := oauth2.NewClient(ctx, ts)

		//TODO: autopaginate turned on.
		//TODO: add support for alternative api endpoints "scm_github_api_endpoint"
		g.client = github.NewClient(tc)
	}

	return nil
}


// configure method will retrieve payload data from Scm using authenticated client.
// MUST set options.IsPullRequest
// RETURNS ScmPayload
func (g *scmGithub) RetrievePayload() (*ScmPayload, error) {
	if !config.IsSet("scm_pull_request") {
		log.Print("This is not a pull request. No automatic continuous deployment processing required. Continuous Integration testing will continue.")
		g.options.IsPullRequest = false

		return &ScmPayload{
			Head: &ScmCommitInfo{
				Sha: config.GetString("scm_sha"),
				Ref: config.GetString("scm_branch"),
				Repo: &ScmRepoInfo{
					CloneUrl: config.GetString("scm_clone_url"),
					Name: config.GetString("scm_repo_name"),
					FullName: config.GetString("scm_repo_full_name"),
				},
			},
		}, nil
		//make this as similar to a pull request as possible
	} else {
		g.options.IsPullRequest = true
		ctx := context.Background()
		parts := strings.Split(config.GetString("scm_repo_full_name"), "/")
		pr, _, err := g.client.PullRequests.Get(ctx, parts[0],parts[1], config.GetInt("scm_pull_request"))

		if(err != nil){
			return nil, errors.ScmAuthenticationFailed(fmt.Sprintf("Could not retrieve pull request from Github: %s", err))
		}

		return &ScmPayload{
			Title: pr.GetTitle(),
			Head: &ScmCommitInfo{
				Sha: pr.Head.GetSHA(),
				Ref: pr.Head.GetRef(),
				Repo: &ScmRepoInfo{
					CloneUrl: pr.Head.Repo.GetCloneURL(),
					Name: pr.Head.Repo.GetName(),
					FullName: pr.Head.Repo.GetFullName(),
				},
			},
			Base: &ScmCommitInfo{
				Sha: pr.Base.GetSHA(),
				Ref: pr.Base.GetRef(),
				Repo: &ScmRepoInfo{
					CloneUrl: pr.Base.Repo.GetCloneURL(),
					Name: pr.Base.Repo.GetName(),
					FullName: pr.Base.Repo.GetFullName(),
				},
			},
		}, nil
	}
}


// all capsule CD processing will be kicked off via a payload. In Github's case, the payload is the webhook data.
// should check if the pull request opener even has permissions to create a release.
// all sources should process the payload by downloading a git repository that contains the master branch merged with the test branch
// MUST set options.GitLocalPath
// MUST set options.GitLocalBranch
// MUST set options.GitHeadInfo
// REQUIRES options.GitParentPath
func (g *scmGithub) ProcessPushPayload(payload *ScmPayload) error {
	//set the processed head info
	g.options.GitHeadInfo = payload.Head
	err := g.options.GitHeadInfo.Validate()
	if(err != nil){
		return err
	}

	if(config.IsSet("scm_github_access_token")){
		// set the remote url, with embedded token
		u, err := url.Parse(g.options.GitHeadInfo.Repo.CloneUrl)
		if err != nil {
			return err
		}

		u.User = url.UserPassword(config.GetString("scm_github_access_token"), "")
		g.options.GitRemote  = u.String()
	} else {
		g.options.GitRemote = g.options.GitHeadInfo.Repo.CloneUrl
	}

	g.options.GitLocalBranch = g.options.GitHeadInfo.Ref

	// clone the merged branch
	// https://sethvargo.com/checkout-a-github-pull-request/
	// https://coderwall.com/p/z5rkga/github-checkout-a-pull-request-as-a-branch

	gitLocalPath, cerr := utils.GitClone(g.options.GitParentPath, g.options.GitHeadInfo.Repo.Name, g.options.GitRemote)
	if(cerr != nil){return cerr}
	g.options.GitLocalPath = gitLocalPath

	return utils.GitCheckout(g.options.GitLocalPath, g.options.GitHeadInfo.Ref)
}

// all capsule CD processing will be kicked off via a payload. In Github's case, the payload is the pull request data.
// should check if the pull request opener even has permissions to create a release.
// all sources should process the payload by downloading a git repository that contains the master branch merged with the test branch
// MUST set options.GitLocalPath
// MUST set options.GitLocalBranch
// MUST set options.GitBaseInfo
// MUST set options.GitHeadInfo
// REQUIRES client
// REQUIRES options.GitParentPath
func (g *scmGithub) ProcessPullRequestPayload(payload *ScmPayload) error {
	return nil
}

func (g *scmGithub) Publish() error {
	return nil
}

func (g *scmGithub) Notify() error {
	return nil
}

func (g *scmGithub) Options() *ScmOptions {
	log.Print("ORINT THE PARENT PATH", g.options)
	return g.options
}